apiserver/facades/client/modelmanager/listmodelsinfo_test.go

```diff
@@ -0,0 +1,177 @@
+)
+
+type ListModelsWithInfoSuite struct {
+	gitjujutesting.IsolationSuite
+
+	st *mockState
+
+	authoriser apiservertesting.FakeAuthorizer
+	adminUser  names.UserTag
+
+	api *modelmanager.ModelManagerAPI
+}
+
+var _ = gc.Suite(&ListModelsWithInfoSuite{})
+
+func (s *ListModelsWithInfoSuite) SetUpTest(c *gc.C) {
+	s.IsolationSuite.SetUpTest(c)
+
+	adminUser := "admin"
+	s.adminUser = names.NewUserTag(adminUser)
+
+	s.st = &mockState{
+		model: s.createModel(c, s.adminUser),
+	}
+	s.st.modelDetailsForUser = func() ([]state.ModelDetails, error) {
+		return []state.ModelDetails{s.st.model.getModelDetails()}, s.st.NextErr()
+	}
+
+	s.authoriser = apiservertesting.FakeAuthorizer{
+		Tag: s.adminUser,
+	}
+	api, err := modelmanager.NewModelManagerAPI(s.st, &mockState{}, nil, s.authoriser, s.st.model)
+	c.Assert(err, jc.ErrorIsNil)
+	s.api = api
+}
+
+func (s *ListModelsWithInfoSuite) createModel(c *gc.C, user names.UserTag) *mockModel {
+	attrs := dummy.SampleConfig()
+	attrs["agent-version"] = jujuversion.Current.String()
+	cfg, err := config.New(config.UseDefaults, attrs)
+	c.Assert(err, jc.ErrorIsNil)
+
+	return &mockModel{
+		owner: user,
+		cfg:   cfg,
+	}
+}
+
+func (s *ListModelsWithInfoSuite) setAPIUser(c *gc.C, user names.UserTag) {
+	s.authoriser.Tag = user
+	modelmanager, err := modelmanager.NewModelManagerAPI(s.st, &mockState{}, nil, s.authoriser, s.st.model)
+	c.Assert(err, jc.ErrorIsNil)
+	s.api = modelmanager
+}
+
+func (s *ListModelsWithInfoSuite) TestListModelsWithInfoForSelf(c *gc.C) {
+	user := names.NewUserTag("external@remote")
+	s.setAPIUser(c, user)
+	s.st.model = s.createModel(c, user)
+	result, err := s.api.ListModelsWithInfo(params.Entity{Tag: user.String()})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Results, gc.HasLen, 1)
+}
+
+func (s *ListModelsWithInfoSuite) TestListModelsWithInfoForSelfLocalUser(c *gc.C) {
+	// When the user's credentials cache stores the simple name, but the
+	// api server converts it to a fully qualified name.
+	user := names.NewUserTag("local-user")
+	s.setAPIUser(c, user)
+	s.st.model = s.createModel(c, user)
+	result, err := s.api.ListModelsWithInfo(params.Entity{Tag: user.String()})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Results, gc.HasLen, 1)
+}
+
+func (s *ListModelsWithInfoSuite) TestListModelsWithInfoAdminSelf(c *gc.C) {
+	result, err := s.api.ListModelsWithInfo(params.Entity{Tag: s.adminUser.String()})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Results, gc.HasLen, 1)
+
+	model := result.Results[0].Result
+	c.Check(model.Name, gc.Equals, "only")
+	c.Check(model.OwnerTag, gc.Equals, s.adminUser.String())
+}
+
+func (s *ListModelsWithInfoSuite) TestListModelsWithInfoAdminListsOther(c *gc.C) {
+	otherTag := names.NewUserTag("someotheruser")
+	s.st.model = s.createModel(c, otherTag)
+	result, err := s.api.ListModelsWithInfo(params.Entity{Tag: s.adminUser.String()})
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(result.Results, gc.HasLen, 1)
```

Comment:
Shouldn't this actually be returning 2 models? Since we have the admin model that was returned around line 106 and then we should have this new model.

Maybe its just that we need to update the mock to more accurately represent filtering, etc?
Or maybe we can strip out a lot of these tests, just test what the mock is accurately testing here, and then do the rest of these tests as unit tests on the state objects.